gensim/models/nmf.py

```diff
@@ -548,10 +556,18 @@ def _setup(self, corpus):
 
         if isinstance(corpus, scipy.sparse.csc.csc_matrix):
             first_doc = corpus.getcol(0)
-        else:
-            first_doc_it = itertools.tee(corpus, 1)
-            first_doc = next(first_doc_it[0])
+        elif hasattr(corpus, "__iter__"):  # check if corpus is iterable
+            if corpus.__iter__() is corpus:  # check if corpus is an iterator
+                if self.passes > 1:
+                    raise ValueError("Corpus is an iterator, only `passes=1` is possible")
+
+                first_doc_it, corpus = itertools.tee(corpus, 2)
+                first_doc = next(iter(first_doc_it))
+            else:
+                first_doc = next(iter(corpus))
+
             first_doc = matutils.corpus2csc([first_doc], len(self.id2word))
+
         self.w_std = np.sqrt(first_doc.mean() / (self.num_tokens * self.num_topics))
```

Comment:
What happens when `hasattr(corpus, "__iter__")` is `False`? Isn't `first_doc` left uninitialized => error on this line?